### PR TITLE
add incontiguous support for cutensor functions

### PR DESCRIPTION
**Pull request by jemiryguo**
Related issue: https://github.com/cupy/cupy/issues/8133

Related PR: https://github.com/cupy/cupy/pull/8136

### CODE REVIEW

**Comment by wxj6000** [comment]
Nice work!

File: cupyx/cutensor.pyx [hunk 1]
```diff
@@ -243,7 +243,7 @@ cpdef TensorDescriptor create_tensor_descriptor(_ndarray_base a):
     """
     handle = _get_handle()
     key = (handle.ptr, a.dtype, tuple(a.shape), tuple(a.strides))
```
Comment (leofang) [inline cupyx/cutensor.pyx:245]: If we want to relax the alignment requirement, we need to add it as part of the cache key. For different alignments, the tensor descriptor cannot be reused.

**Review by leofang — CHANGES_REQUESTED** [review]
In order to relax the alignment requirement, we need to make it a part of cache key for all objects that we cache internally, as long as creation of such objects require the presence of valid tensor descriptors, either directly or indirectly. This includes
- tensor descriptors,
- all operators (elementwise/reduction/contraction/etc)
- plans built on top of them

They need to be re-created as needed, if the new alignment mismatches with that used at object creation time.

File: cupyx/cutensor.pyx (hunk 1 above)
Comment (jemiryguo) [reply cupyx/cutensor.pyx:245]: I got it, I will change that

Comment (jemiryguo) [reply cupyx/cutensor.pyx:245]: I found that we only need to change the key to cache descripters. For other cached object, desc.ptr is used in keys thus naturally be recreated.

Comment (leofang) [reply cupyx/cutensor.pyx]: I see, it's a good observation, so we're using the tensor descriptor pointer as a proxy to propagate the alignment requirement. As long as we don't have dangling pointers, this should work fine.

**Comment by leofang** [comment]
@jemiryguo could you check the linter error please?

File: tests/cupyx_tests/test_cutensor.py [hunk 2]
```diff
@@ -337,3 +337,133 @@ def test_contraction(self):
...
+        r_dtype = dtype
+        if dtype == numpy.complex64:
+            r_dtype = numpy.float32
+        elif dtype == numpy.complex128:
+            r_dtype = numpy.float64
+        a = testing.shaped_random(shape, cupy, dtype=r_dtype, scale=1)
+        if dtype.char in 'FD':
+            a = a + 1j * \
+                testing.shaped_random(shape, cupy, dtype=r_dtype, scale=1)
+        return a
```
Comment (leofang) [inline tests/cupyx_tests/test_cutensor.py:362]: This should not be necessary, `testing.shaped_random` can generate complex random arrays too, so just use it everywhere.

File: tests/cupyx_tests/test_cutensor.py [hunk 3]
```diff
@@ -337,3 +337,133 @@ def test_contraction(self):
                              self.c, mode_c)
         cupy.testing.assert_allclose(self.c, self.c_ref,
                                      rtol=self.tol, atol=self.tol)
+
+
+@testing.parameterize(*testing.product({
+    'dtype_char': ['e', 'f', 'd', 'F', 'D'],
+    'shape': [(30, 40, 30, 35)],
+    'alpha': [0.5, 1.0],
+    'beta': [0.0, 1.0],
```
Comment (leofang) [inline tests/cupyx_tests/test_cutensor.py:346]: Let's also test different orders 
```suggestion
    'beta': [0.0, 1.0],
    'order': ('C', 'F'),
```
(need to pass `self.order` to `shaped_random`).

**Review by leofang — CHANGES_REQUESTED** [review]
Thanks, @jemiryguo. The tests look good overall. I suggested two minor changes. Once they're in I'll kick off the CI.

**Review by leofang — APPROVED** [review]

**Comment by leofang** [comment]
@asi1024 I think this is ready 🙂 btw I added the backport label to unblock @jemiryguo and his collaborators sooner, as they have memory-demanding applications.

**Review by asi1024 — APPROVED** [review]

**Comment by asi1024** [comment]
LGTM. Thanks!

**Pull request merged**